consensus/src/pending_votes.rs

```diff
@@ -56,15 +62,120 @@ pub enum VoteReceptionResult {
+    ) {
+        self.partial_2chain_tc.add(author, timeout, signature);
+        self.timeout_reason.entry(author).or_insert(reason);
+    }
+
+    pub(super) fn partial_2chain_tc_mut(&mut self) -> &mut TwoChainTimeoutWithPartialSignatures {
+        &mut self.partial_2chain_tc
+    }
+
+    pub(super) fn partial_2chain_tc(&self) -> &TwoChainTimeoutWithPartialSignatures {
+        &self.partial_2chain_tc
+    }
+
+    fn aggregated_timeout_reason(&self, verifier: &ValidatorVerifier) -> RoundTimeoutReason {
+        let mut reason_voting_power: HashMap<RoundTimeoutReason, u128> = HashMap::new();
+        let mut missing_batch_authors: HashMap<Author, u128> = HashMap::new();
+        let ordered_authors = verifier.get_ordered_account_addresses();
+        for (author, reason) in &self.timeout_reason {
+            // To aggregate the reason, we only care about the variant type itself and
+            // exclude any data within the variants.
+            let reason_key = match reason {
+                reason @ RoundTimeoutReason::Unknown
+                | reason @ RoundTimeoutReason::ProposalNotReceived
+                | reason @ RoundTimeoutReason::NoQC => reason.clone(),
+                RoundTimeoutReason::PayloadUnavailable { missing_authors } => {
+                    for missing_idx in missing_authors.iter_ones() {
+                        if let Some(missing_author) = ordered_authors.get(missing_idx) {
+                            *missing_batch_authors.entry(*missing_author).or_default() +=
+                                verifier.get_voting_power(missing_author).unwrap() as u128;
+                        }
+                    }
+                    RoundTimeoutReason::PayloadUnavailable {
+                        missing_authors: BitVec::with_num_bits(verifier.len() as u16),
+                    }
+                },
+            };
+            *reason_voting_power.entry(reason_key).or_default() +=
+                verifier.get_voting_power(&author).unwrap_or_default() as u128;
+        }
+        let aggregated_reason = reason_voting_power
+            .into_iter()
+            .max_by_key(|(_, voting_power)| *voting_power)
+            .and_then(|(reason, voting_power)| {
+                if verifier
+                    .check_aggregated_voting_power(voting_power, false)
+                    .is_ok()
+                {
+                    // If the aggregated reason is due to unavailable payload, we will compute the
+                    // aggregated missing authors bitvec counting batch authors that have been reported
+                    // missing by minority peers.
+                    if matches!(reason, RoundTimeoutReason::PayloadUnavailable { .. }) {
+                        let mut aggregated_bitvec = BitVec::with_num_bits(verifier.len() as u16);
+                        let address_to_index = verifier.address_to_validator_index();
+                        for (author, voting_power) in missing_batch_authors {
```

Comment:
doesn't look like this needs to be Author -> u128 mapping, it can stay as index -> u128 mapping?